### PR TITLE
bpf: Define new CILIUM_NOTIFY_TRACE perf ring buffer msg type

### PR DESCRIPTION
**Pull request by rlenglet**

### CODE REVIEW

File: bpf/lib/drop.h [hunk 1]
```diff
@@ -36,7 +36,7 @@
 #ifdef DROP_NOTIFY
 __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_DROP_NOTIFY) int __send_drop_notify(struct __sk_buff *skb)
 {
-	uint64_t skb_len = skb->len, cap_len = min(64ULL, skb_len);
+	uint64_t skb_len = skb->len, cap_len = min(TRACE_PAYLOAD_LEN, skb_len);
```
Comment (tgraf) [inline bpf/lib/drop.h]: 👍

File: bpf/bpf_lxc.c [hunk 2]
```diff
@@ -265,7 +266,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 		if (ret != TC_ACT_OK)
 			return ret;
 
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_PROXY`

File: bpf/bpf_lxc.c [hunk 3]
```diff
@@ -368,7 +369,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 			return ret;
 
 #ifndef POLICY_ENFORCEMENT
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_HOST`

File: bpf/bpf_lxc.c [hunk 4]
```diff
@@ -395,7 +396,7 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 
 #ifndef POLICY_ENFORCEMENT
 	/* No policy, pass directly down to stack */
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, 0);
+	send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, 0);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_STACK`

File: bpf/bpf_lxc.c [hunk 5]
```diff
@@ -587,7 +588,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 		if (ret != TC_ACT_OK)
 			return ret;
 
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_PROXY`

File: bpf/bpf_lxc.c [hunk 6]
```diff
@@ -677,7 +678,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 			return ret;
 
 #ifndef POLICY_ENFORCEMENT
-		cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, HOST_IFINDEX);
+		send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, HOST_IFINDEX);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_HOST`

File: bpf/bpf_lxc.c [hunk 7]
```diff
@@ -706,7 +707,7 @@ static inline int handle_ipv4(struct __sk_buff *skb)
 
 #ifndef POLICY_ENFORCEMENT
 	/* No policy, pass directly down to stack */
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, 0);
+	send_trace_notify(skb, TRACE_FROM_LXC, SECLABEL, 0, 0, 0);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: `TRACE_TO_STACK`

File: bpf/bpf_lxc.c [hunk 8]
```diff
@@ -1024,9 +1025,7 @@ __section_tail(CILIUM_MAP_POLICY, LXC_ID) int handle_policy(struct __sk_buff *sk
 		return send_drop_notify(skb, src_label, SECLABEL, LXC_ID,
 					ifindex, ret, TC_ACT_SHOT);
 
-	ifindex = skb->cb[CB_IFINDEX];
-
-	cilium_dbg_capture(skb, DBG_CAPTURE_DELIVERY, ifindex);
+	send_trace_notify(skb, TRACE_TO_LXC, src_label, SECLABEL, LXC_ID, ifindex);
```
Comment (tgraf) [inline bpf/bpf_lxc.c]: 👍

File: daemon/main.go [hunk 9]
```diff
@@ -447,6 +447,7 @@ func initEnv() {
 	}
 
 	config.Opts.Set(endpoint.OptionDropNotify, true)
+	config.Opts.Set(endpoint.OptionTraceNotify, true)
```
Comment (tgraf) [inline daemon/main.go:446]: 👍

**Review by tgraf — CHANGES_REQUESTED** [review]

File: cilium/cmd/endpoint_config.go [hunk 10]
```diff
@@ -30,7 +30,7 @@ var listOptions bool
 var endpointConfigCmd = &cobra.Command{
 	Use:     "config <endpoint id> [<option>=(enable|disable) ...]",
 	Short:   "View & modify endpoint configuration",
-	Example: "endpoint config 5421 DropNotification=disable",
+	Example: "endpoint config 5421 DropNotification=false TraceNotification=false",
```
Comment (aanm) [inline cilium/cmd/endpoint_config.go:33]: As a user: What's the difference between the Trace and the Debug?

Comment (tgraf) [reply cilium/cmd/endpoint_config.go:33]: Should be simple: Debug should be a developer only flag. We should remove it from the examples. Trace is for end users as well. The overhead of debug will be higher than the trace.

Comment (rlenglet) [reply cilium/cmd/endpoint_config.go:33]: I think the naming is confusing. The `CILIUM_NOTIFY_DBG_CAPTURE` messages are sent if `ENABLE_TRACE` is defined. That is currently always defined, and it's defined independently from `DEBUG`. `ENABLE_TRACE` can be undefined if the `--disable-trace` is given to Cilium. However, that option is used nowhere:
```
$ grep -r disable-trace *
daemon/main.go:		"disable-trace", false, "Disable ability to trace datapath actions")
```
Therefore, we currently always send the `CILIUM_NOTIFY_DBG_CAPTURE` messages, which I believe is a bug.

As @tgraf mentioned, the new TraceNotification is to enable a new set of messages, which are not only for debugging, but are as useful to users as drop messages.

I propose to remove the `ENABLE_TRACE` define, and use `DEBUG` instead to enable the `CILIUM_NOTIFY_DBG_CAPTURE` messages. I'll also remove the currently unused `--disable-trace` option.

Comment (rlenglet) [reply cilium/cmd/endpoint_config.go:33]: Pushed a new commit to replace `ENABLE_TRACE` with `DEBUG` and remove `disable-trace`.

File: pkg/bpfdebug/trace.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+// DumpInfo prints a summary of the trace messages.
+func (n *TraceNotify) DumpInfo(data []byte) {
+	fmt.Printf("xx forward (%s) to endpoint %d, identity %d->%d: %s\n",
+		obsPoint(n.ObsPoint), n.DstID, n.SrcLabel, n.DstLabel,
+		GetConnectionSummary(data[TraceNotifyLen:]))
+}
+
+// Dump prints the trace notification in human readable form
```
[the remark this answers is not in the document]
Comment (rlenglet) [reply pkg/bpfdebug/trace.go]: Done.

[the remark this answers is not in the document]
Comment (rlenglet) [reply pkg/bpfdebug/trace.go]: Done.

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Could it be better to include both source and destination in the observation point names, e.g. `TRACE_LXC_TO_PROXY`?

**Comment by rlenglet** [comment]
tests/10-proxy.sh has curl consistently timeout in the "Testing with Policy=ingress, Conntrack=false, Service=none". egress doesn't timeout. I'm investigating.

File: bpf/bpf_lxc.c (hunk 2 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

File: bpf/bpf_lxc.c (hunk 3 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

File: bpf/bpf_lxc.c (hunk 4 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

File: bpf/bpf_lxc.c (hunk 5 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

File: bpf/bpf_lxc.c (hunk 6 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

File: bpf/bpf_lxc.c (hunk 7 above)
Comment (rlenglet) [reply bpf/bpf_lxc.c]: Done.

**Comment by rlenglet** [comment]
I fixed the tests and addressed the comments. Please review.

**Review by aanm — APPROVED** [review]

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
Great work!

**Pull request merged**